lib/evervault/crypto/client.rb

```diff
@@ -42,26 +45,59 @@ def encrypt(data)
         iv = cipher.random_iv
         cipher.iv = iv
 
-        if (@curve == 'prime256v1')
-          cipher.auth_data = Base64.strict_decode64(@team_key)
-        end
+        cipher.auth_data = Base64.strict_decode64(@team_key)
+
+        metadata = generate_metadata(role)
+        metadata_offset = [metadata.length].pack("v") # 'v' specifies 16-bit unsigned little-endian
+        payload = metadata_offset + metadata + data_to_encrypt.to_s
 
-        encrypted_data = cipher.update(data_to_encrypt.to_s) + cipher.final + cipher.auth_tag
+        encrypted_data = cipher.update(payload.to_s) + cipher.final + cipher.auth_tag
 
         ephemeral_key_compressed_string = @ephemeral_public_key.to_octet_string(:compressed)
 
         format(header_type(data_to_encrypt), Base64.strict_encode64(iv), Base64.strict_encode64(ephemeral_key_compressed_string), Base64.strict_encode64(encrypted_data))
       end
 
-      private def traverse_and_encrypt(data)
+      def generate_metadata(role)
+        buffer = []
+
+        # Binary representation of a fixed map with 2 or 3 items, followed by the key-value pairs
+        buffer.push(0x80 | (role.nil? ? 2 : 3))
+
+        if role
+          # Binary representation for a fixed string of length 2, followed by `dr` (for "data role")
+          buffer.push(0xA2)
+          buffer.push(*"dr".bytes)
+          # Binary representation for a fixed string of role name length, followed by the role name itself
+          buffer.push(0xA0 | role.length)
+          buffer.push(*role.bytes)
+        end
+
+        # Binary representation for a fixed string of length 2, followed by `eo` (for "encryption origin")
+        buffer.push(0xA2)
+        buffer.push(*"eo".bytes)
+        # Binary representation for the integer 8 (Ruby SDK)
+        buffer.push(8)
+
+        # Binary representation for a fixed string of length 2, followed by `et` (for "encryption timestamp")
+        buffer.push(0xA2)
+        buffer.push(*"et".bytes)
+        # Binary representation for a 4-byte unsigned integer (uint 32), followed by the epoch time
+        buffer.push(0xCE)
+        buffer.push(*[Time.now.to_i].pack(">I").bytes)
```

Comment:
It seems that `>I` isn't a valid directive here. Think it can just be `I`.

Calling `Evervault.encrypt` is giving the following wanring: 
```
<internal:pack>:8: warning: unknown pack directive '>' in '>I'
```

Reply:
You're absolutely correct, that was the python expression but it's different in Ruby. I've changed to `I!<` as that's the expression for big endian as noted here: https://apidock.com/ruby/Array/pack